### DISCUSSION TITLE
Support TrustedTypes / Sanitization

### DISCUSSION THREAD

**Discussion by raphael-inglin-ergon** [sibiraj-s/ngx-editor#480] [Ideas]
Hi

For our app, we require trusted types via our CSP policy (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/trusted-types). Ngx-Editor currently has a few direct assignments to innerHtml without sanitization or trustedTypes, which violates our policy.

I would like to add an optional flag to turn on sanitization (via DOMPurify) for these assignments while keeping the current behavior if not active. This would allow me to use the "dompurify" trusted types and satisfy the policy.

I'd be happy to provide a PR. Any objections or ideas?

**Comment by sibiraj-s** [comment]
> I'd be happy to provide a PR. Any objections or ideas?

Hi, can you share some ideas here on how you would like to implement it? before raise a PR, typically we don't handle html, prosemirror does don't want to to drag by directly jumping into the PR.

you can explore the code and let me know if you still think the implementation is possible?

**Reply by raphael-inglin-ergon** [reply]
Hi. Thanks for the hints. I checked the code again, these are the places that would need a change:

1. EditorCommands.ts -> insertHtml & parsers.ts -> toDoc
Here, there is an assignment to innerHtml which is picked up by the trustedTypes. Since the created element is only passed to prosemirror and not rendered, it should be safe, but the browser doesnt know that. So we could
- Add a ngx-editor trustedType which does a noop and guarantees it is safe, but does not change the html (similar to how angular or DOMPurify do it internally)
- OR Actually sanitize it using e.g. DOMPurify, which could potentially change the html, is definitely safe, but might be considered a breaking change

2. Icons [innerHtml]
Here the sanitizeHtml.ts pipe is used to render the icons from the default or user config. It uses the Angular `bypassSecurityTrustHtml` which does no sanitation, but just tells us it is safe (via angular#unsafe-bypass) . Since we cannot guarantee where the html comes from, here unsafe html could be rendered.
- Doing actual sanitizaion e.g. DOMPurify would be the safest way, but could be considered a breaking change

There might be a way to support the sanitization alongside the current approach with configuration passed in from the user outside to make sure this is not a breaking change, but it would be a bit cumbersome. 

Let me know what you think about these suggestions.

(EDIT: some background info on trusted-types: https://web.dev/trusted-types/)

**Reply by sibiraj-s** [reply]
Let me check this out, but in general I am hesitant to add DOMPurify to the system, 

1. its potentially breaking
2. adds one more dependency to the lib
3. Might sit between prose-mirror and user and open up new set of problems to deal with as the content generated and output might not match exactly

its always better if sanitization is done by the user. I am fine by adding some strict types to make sure user passes sanitized content. Thoughts?

**Reply by raphael-inglin-ergon** [reply]
Hi. I understand your concerns. I did some more experiments and it seems to be possible to just allow passing a `TrustedHTML | string` (from @types/trusted-types) wherever strings are used for HTML at the moment and adjust some conditions slightly. That way the existing logic is unaffected and the user can pass safe HTML or simple strings as before.

Would that work for you?

**Reply by raphael-inglin-ergon** [reply]
I took the liberty to create a small PR: https://github.com/sibiraj-s/ngx-editor/pull/481 . I hope it aligns more or less with what you had in mind. Otherwise please let me know.

**Reply by sibiraj-s** [reply]
Thanks. I'll check that out. it is still a breaking change correct?

**Reply by raphael-inglin-ergon** [reply]
It should not be a breaking change unless I missed something somewhere. No changes are made to the incoming data, it only extends the existing code to accept TrustedHTML values and handle them correctly. If strings are passed, they should be handled exactly like before.

**Reply by sibiraj-s** [reply]
Ah got it. I didn't yet check the PR. I'll check shortly. Thanks for the explanation

**Reply by sibiraj-s** [reply]
Merged the PR. Will release soon. Thanks

**Comment by vonox7** [comment]
Fyi, to use trusted types in ngx-editor, the Content-Security-Policy header must have `trusted-types angular#unsafe-bypass`. 

With only `trusted-types angular` you would get the following errors:
```
language.service.ts:267 Refused to create a TrustedTypePolicy named 'angular#unsafe-bypass' because it violates the following Content Security Policy directive: "trusted-types angular 'allow-duplicates'".

This document requires 'TrustedHTML' assignment.Understand this errorAI
language.service.ts:267 TypeError: Failed to set the 'innerHTML' property on 'Element': This document requires 'TrustedHTML' assignment.
    at EmulatedEncapsulationDomRenderer2.setProperty (platform-browser.mjs:645:17)
    at BaseAnimationRenderer.setProperty (browser.mjs:4632:27)
    at elementPropertyInternal (core.mjs:11711:18)
    at ɵɵproperty (core.mjs:21289:9)
    at ToggleCommandComponent_Template (ngx-editor.mjs:1342:408)
    at executeTemplate (core.mjs:11268:9)
    at refreshView (core.mjs:12791:13)
    at detectChangesInView$1 (core.mjs:13015:9)
    at detectChangesInViewIfAttached (core.mjs:12978:5)
    at detectChangesInComponent (core.mjs:12967:5)
```
